qt/widgets/instrumentview/src/Shape2D.cpp

```diff
@@ -620,6 +623,497 @@ std::string Shape2DRing::saveToProject() const {
+    x_min = std::min(cos(m_startAngle), cos(m_endAngle));
+    x_min = std::min(x_min * m_innerRadius, x_min * m_outerRadius);
+  }
+
+  if ((m_startAngle <= 3 * M_PI / 2 && m_endAngle >= 3 * M_PI / 2) ||
+      (m_startAngle > m_endAngle &&
+       !(m_startAngle >= 3 * M_PI / 2 && m_endAngle <= 3 * M_PI / 2))) {
+    y_min = -m_outerRadius;
+  } else {
+    y_min = std::min(sin(m_startAngle), sin(m_endAngle));
+    y_min = std::min(y_min * m_innerRadius, y_min * m_outerRadius);
+  }
+
+  if (m_startAngle > m_endAngle) {
+    x_max = m_outerRadius;
+  } else {
+    x_max = std::max(cos(m_startAngle), cos(m_endAngle));
+    x_max = std::max(x_max * m_innerRadius, x_max * m_outerRadius);
+  }
+
+  QPointF bottom_left(x_min, y_min);
+  QPointF top_right(x_max, y_max);
+
+  bottom_left += m_center;
+  top_right += m_center;
+
+  m_boundingRect = RectF(bottom_left, top_right);
+}
+
+QPointF Shape2DSector::getShapeControlPoint(size_t i) const {
+  double halfAngle =
+      m_startAngle < m_endAngle
+          ? fmod((m_startAngle + m_endAngle) / 2., 2 * M_PI)
+          : fmod((m_startAngle + m_endAngle + 2 * M_PI) / 2, 2 * M_PI);
+  double halfLength = (m_outerRadius + m_innerRadius) / 2;
+
+  switch (i) {
+  case 0:
+    return QPointF(m_center.x() + cos(halfAngle) * m_outerRadius,
+                   m_center.y() + sin(halfAngle) * m_outerRadius);
+  case 1:
+    return QPointF(m_center.x() + cos(halfAngle) * m_innerRadius,
+                   m_center.y() + sin(halfAngle) * m_innerRadius);
+  case 2:
+    return QPointF(m_center.x() + cos(m_startAngle) * halfLength,
+                   m_center.y() + sin(m_startAngle) * halfLength);
+  case 3:
+    return QPointF(m_center.x() + cos(m_endAngle) * halfLength,
+                   m_center.y() + sin(m_endAngle) * halfLength);
+  }
+  return QPointF();
+}
+
+void Shape2DSector::setShapeControlPoint(size_t i, const QPointF &pos) {
+  QPointF to_center = pos - m_center;
+  double newAngle;
+
+  switch (i) {
+  case 0:
+    m_outerRadius = distanceBetween(to_center, QPointF(0, 0));
+    if (m_outerRadius < m_innerRadius) {
+      m_outerRadius = 1.01 * m_innerRadius;
+    }
+    break;
+  case 1:
+    m_innerRadius = distanceBetween(to_center, QPointF(0, 0));
+    if (m_outerRadius < m_innerRadius) {
+      m_innerRadius = 0.99 * m_outerRadius;
+    }
+    break;
+  case 2:
+    newAngle = atan2(to_center.y(), to_center.x());
+    if (newAngle < 0)
+      newAngle += 2 * M_PI;
+
+    // conditions to prevent the startAngle from going over the endAngle
+    if ((m_startAngle < m_endAngle && newAngle >= m_endAngle &&
+         std::abs(newAngle - m_startAngle) < 1) ||
+        (newAngle < m_endAngle && m_startAngle < m_endAngle &&
+         std::abs(newAngle - m_startAngle) > 5.28 && newAngle < m_startAngle) ||
+        (newAngle > m_endAngle && m_startAngle > m_endAngle &&
+         std::abs(newAngle - m_startAngle) > 5.28 && newAngle < m_startAngle)) {
+
+      newAngle = m_endAngle - 0.00001;
```

Comment:
This is not the same as in line 991. Can you make a global constexpr for this tolerance just as for the epsillon?